test/tests/functional/pbs_hook_execjob_prologue.py

```diff
@@ -0,0 +1,234 @@
+                                 n=100, max_attempts=10, interval=2)
+        self.assertTrue(rv)
+
+
+    def test_prologue_internal_error_no_fail_action(self):
+        """
+        Test a prologue hook with an internal error and no fail_action.
+        """
+        hook_name = "prologue_exception"
+        # multi-line string format using tuple
+        hook_body = ("import pbs\n"
+                     "e = pbs.event()\n"
+                     "x\n")
+
+        attr = {'event': 'execjob_prologue',
+                'enabled': 'True'}
+        self.server.create_import_hook(hook_name, attr, hook_body)
+
+        j = Job(TEST_USER)
+        attr = {'Resource_List.select': 'vnode={}'.format(self.hostA),
+                'Resource_List.walltime': 30}
+        j.set_attributes(attr)
+        j.set_sleep_time(10)
+        jid = self.server.submit(j)
+
+        self.server.expect(NODE, {'state': 'free'}, id=self.hostA, interval=4)
+
+
+    def test_prologue_internal_error_offline_vnodes(self):
+        """
+        Test a prologue hook with an internal error and
+        fail_action=offline_vnodes.
+        """
+        attr = {'resources_available.mem': '2gb',
+                'resources_available.ncpus': '1'}
+        self.server.create_vnodes(self.hostC, attr, 2, self.momC, delall=True)
+        self.server.manager(MGR_CMD_SET, NODE, attr, id=self.hostC)
+
+        hook_name = "prologue_exception"
+        # multi-line string format using tuple
+        hook_body = ("import pbs\n"
+                     "e = pbs.event()\n"
+                     "x\n")
+        attr = {'event': 'execjob_prologue',
+                'enabled': 'True',
+                'fail_action': 'offline_vnodes'}
+        self.server.create_import_hook(hook_name, attr, hook_body)
+
+        j = Job(TEST_USER)
+        attr = {'Resource_List.select': 'vnode={}[0]'.format(self.hostC),
+                'Resource_List.walltime': 30}
+        j.set_attributes(attr)
+        j.set_sleep_time(10)
+        jid = self.server.submit(j)
+
+        attr = {
+            'state': 'offline',
+            'comment': "offlined by hook '%s' due to hook error" % (hook_name,)
+        }
+        self.server.expect(VNODE, attr, id=self.hostC, op=MATCH,
+                           max_attempts=10, interval=2)
+        self.server.expect(VNODE, attr, id='{}[0]'.format(self.hostC),
+                           op=MATCH, max_attempts=10, interval=2)
+        self.server.expect(VNODE, attr, id='{}[1]'.format(self.hostC),
+                           op=MATCH, max_attempts=10, interval=2)
+
+        # revert momC
+        self.server.manager(MGR_CMD_SET, NODE, {'state': 'free'},
+                            id=self.hostC)
+        self.server.manager(MGR_CMD_SET, NODE, {'state': 'free'},
+                            id='{}[0]'.format(self.hostC))
+        self.server.manager(MGR_CMD_SET, NODE, {'state': 'free'},
+                            id='{}[1]'.format(self.hostC))
```

Comment:
Instead of setting the state to free, I'd rather you do {'state': (DECR, 'offline)}.  This allows the server to correctly set the state.